### PR TITLE
executor: support chunk for IndexLookupReader.

### PR DESCRIPTION
**Pull request by coocood**
And decouple indexWorker and tableWorker from index lookup executor.

### CODE REVIEW

**Comment by coocood** [comment]
@zz-jason @tiancaiamao PTAL

File: executor/distsql.go [hunk 1]
```diff
@@ -714,75 +704,59 @@ func (e *IndexLookUpExecutor) indexRangesToKVRanges() ([]kv.KeyRange, error) {
...
-
-func (e *IndexLookUpExecutor) buildTableTasks(handles []int64) []*lookupTableTask {
-	// Build tasks with increasing batch size.
-	var taskSizes []int
-	for remained := len(handles); remained > 0; e.batchSize *= 2 {
-		if e.batchSize > e.ctx.GetSessionVars().IndexLookupSize {
-			e.batchSize = e.ctx.GetSessionVars().IndexLookupSize
+	tps := tableReader.Schema().GetTypes()
+	for {
+		chk := chunk.NewChunk(tps)
```
Comment (zz-jason) [inline executor/distsql.go]:
```go
chk := tableReader.newChunk()
```

File: executor/distsql.go [hunk 2]
```diff
@@ -714,75 +704,59 @@ func (e *IndexLookUpExecutor) indexRangesToKVRanges() ([]kv.KeyRange, error) {
...
+			return
 		}
-		if e.batchSize > remained {
-			taskSizes = append(taskSizes, remained)
+		if chk.NumRows() == 0 {
 			break
 		}
-		taskSizes = append(taskSizes, e.batchSize)
-		remained -= e.batchSize
+		for i := 0; i < chk.NumRows(); i++ {
```
Comment (zz-jason) [inline executor/distsql.go]:
```go
for row := chk.Begin(); row != chk.End(); row = row.Next() {
}
```

File: executor/distsql.go [hunk 3]
```diff
@@ -714,75 +704,59 @@ func (e *IndexLookUpExecutor) indexRangesToKVRanges() ([]kv.KeyRange, error) {
...
-	for {
-		var row Row
-		row, err = tableReader.Next()
-		if err != nil || row == nil {
-			break
-		}
-		task.rows = append(task.rows, row)
+	if w.keepOrder {
+		task.rows = make([]chunk.Row, len(task.handles))
 	}
```
Comment (zz-jason) [inline executor/distsql.go]:
```go
else {
    task.rows = make([]chunk.Row, 0, len(task.handles))
}
```

File: executor/distsql.go [hunk 4]
```diff
@@ -586,99 +553,124 @@ func (e *IndexLookUpExecutor) startIndexWorker(goCtx goctx.Context, kvRanges []k
...
 }
 
-func (worker *tableWorker) close() {
-	worker.wg.Wait()
-}
-
 // Open implements the Executor Open interface.
 func (e *IndexLookUpExecutor) Open(goCtx goctx.Context) error {
 	kvRanges, err := e.indexRangesToKVRanges()
```
Comment (zz-jason) [inline executor/distsql.go]: 1. can we remove `e.indexRangesToKVRanges` and use `indexRangesToKVRanges` directly ?
2. can we merge `e.Open` and `e.open` to `e.Open` ?

File: executor/distsql.go [hunk 5]
```diff
@@ -551,24 +514,28 @@ type IndexLookUpExecutor struct {
...
 
 // indexWorker is used by IndexLookUpExecutor to maintain index lookup background goroutines.
 type indexWorker struct {
-	wg sync.WaitGroup
+	workCh    chan<- *lookupTableTask
+	finished  <-chan struct{}
+	resultCh  chan<- *lookupTableTask
+	keepOrder bool
+
+	// batchSize is for slow startup. It will slowly increase to the max batch size value.
```
Comment (zz-jason) [inline executor/distsql.go]: 1. batchSize's grow speed is not slow, It is doubled every time we call `extractTaskHandles`
2. "be increased to"

File: executor/distsql.go [hunk 6]
```diff
@@ -195,45 +177,26 @@ func indexRangesToKVRanges(tid, idxID int64, ranges []*types.IndexRange) ([]kv.K
...
-		if data == nil {
-			break
+		if chk.NumRows() == 0 {
+			return
 		}
-		handleDatum, err = tablecodec.DecodeColumnValue(data[0].GetRaw(), handleType, nil)
-		if err != nil {
-			return nil, errors.Trace(err)
+		for i := 0; i < chk.NumRows(); i++ {
+			handles = append(handles, chk.GetRow(i).GetInt64(0))
```
Comment (zz-jason) [inline executor/distsql.go]: 1. After the `for` loop, `len(handles)` may be larger than `w.batchSize` and `w.maxBatchSize`
2. Can we add an API for Chunk, like `func (c *Chunk) GetIntVector(colID, begin, end int) []int64`, to get a slice of integers/strings/decimals ...

File: executor/distsql.go [hunk 7]
```diff
@@ -714,75 +704,59 @@ func (e *IndexLookUpExecutor) indexRangesToKVRanges() ([]kv.KeyRange, error) {
...
 	var tableReader Executor
-	tableReader, err = e.dataReaderBuilder.buildTableReaderFromHandles(goCtx, &TableReaderExecutor{
-		baseExecutor: newBaseExecutor(e.schema, e.ctx),
-		table:        e.table,
-		tableID:      e.tableID,
-		dagPB:        e.tableRequest,
-	}, task.handles)
+	tableReader, err = w.buildTblReader(goCtx, task.handles)
 	if err != nil {
 		log.Error(err)
```
Comment (zz-jason) [inline executor/distsql.go]: no need to log it since we will write this error back to `task.doneCh`

File: executor/distsql.go [hunk 8]
```diff
@@ -714,75 +704,59 @@ func (e *IndexLookUpExecutor) indexRangesToKVRanges() ([]kv.KeyRange, error) {
...
 		return
 	}
 	defer terror.Call(tableReader.Close)
```
Comment (zz-jason) [inline executor/distsql.go]: write the error returned by `tableReader.Close()` to `task.doneCh`

File: executor/distsql.go [hunk 9]
```diff
@@ -714,75 +704,59 @@ func (e *IndexLookUpExecutor) indexRangesToKVRanges() ([]kv.KeyRange, error) {
...
-	var taskSizes []int
-	for remained := len(handles); remained > 0; e.batchSize *= 2 {
-		if e.batchSize > e.ctx.GetSessionVars().IndexLookupSize {
-			e.batchSize = e.ctx.GetSessionVars().IndexLookupSize
+	tps := tableReader.Schema().GetTypes()
+	for {
+		chk := chunk.NewChunk(tps)
+		err = tableReader.NextChunk(chk)
+		if err != nil {
+			log.Error(err)
```
Comment (zz-jason) [inline executor/distsql.go]: ditto

File: executor/distsql.go [hunk 10]
```diff
@@ -55,7 +54,7 @@ var LookupTableTaskChannelSize int32 = 50
 // contains the handles in those index keys.
 type lookupTableTask struct {
 	handles []int64
-	rows    []Row
+	rows    []chunk.Row
 	cursor  int
 
 	done   bool
```
Comment (zz-jason) [inline executor/distsql.go]: Can we remove `done` ? Seems it's always `false` ?

File: executor/distsql.go (hunk 6 above)
Comment (coocood) [reply executor/distsql.go]: 1. I know, but I think it's acceptable that the batch size is a little larger.
2. `GetInt64` is fast enough, it only takes less than 2 ns.

File: executor/distsql.go (hunk 7 above)
Comment (coocood) [reply executor/distsql.go]: It's is possible that the error is not received.

File: executor/distsql.go (hunk 8 above)
Comment (coocood) [reply executor/distsql.go]: We only need to log the `Close` error.

File: executor/distsql.go (hunk 4 above)
Comment (coocood) [reply executor/distsql.go]: 2. No, the open method is used by dataReaderBuilder with a datum range.

**Comment by coocood** [comment]
@zz-jason PTAL

File: executor/distsql.go [hunk 11]
```diff
@@ -195,45 +168,26 @@ func indexRangesToKVRanges(tid, idxID int64, ranges []*types.IndexRange) ([]kv.K
...
 		}
-		if data == nil {
-			break
+		if chk.NumRows() == 0 {
+			return
 		}
-		handleDatum, err = tablecodec.DecodeColumnValue(data[0].GetRaw(), handleType, nil)
-		if err != nil {
-			return nil, errors.Trace(err)
+		for i := 0; i < chk.NumRows(); i++ {
```
Comment (tiancaiamao) [inline executor/distsql.go]: Did you consider chunk.MaxCapacity here?

Comment (coocood) [reply executor/distsql.go]: The handle batch size can be greater than chunk.MaxCapacity.
The task gets multiple chunks from multiple regions.

File: executor/distsql.go [hunk 12]
```diff
@@ -586,99 +553,124 @@ func (e *IndexLookUpExecutor) startIndexWorker(goCtx goctx.Context, kvRanges []k
...
 
 // fetchHandles fetches a batch of handles from index data and builds the index lookup tasks.
 // The tasks are sent to workCh to be further processed by tableWorker, and sent to e.resultCh
 // at the same time to keep data ordered.
-func (worker *indexWorker) fetchHandles(e *IndexLookUpExecutor, result distsql.SelectResult, workCh chan<- *lookupTableTask, goCtx goctx.Context, finished <-chan struct{}) {
+func (w *indexWorker) fetchHandles(goCtx goctx.Context, result distsql.SelectResult) {
+	chk := chunk.NewChunk([]*types.FieldType{types.NewFieldType(mysql.TypeLonglong)})
 	for {
-		handles, finish, err := extractHandlesFromIndexResult(result)
+		handles, err := w.extractTaskHandles(chk, result)
```
[the remark this answers is not in the document]
Comment (coocood) [reply executor/distsql.go]: The `SelectResult` will ignore zero bytes response, and return the next non-empty chunk.

File: executor/distsql.go [hunk 13]
```diff
@@ -693,96 +676,74 @@ func (e *IndexLookUpExecutor) open(goCtx goctx.Context, kvRanges []kv.KeyRange)
...
-			break
-		}
-		task.rows = append(task.rows, row)
-	}
-	if e.keepOrder {
-		// Restore the index order.
-		sorter := &rowsSorter{order: task.indexOrder, rows: task.rows, handleIdx: e.handleIdx}
-		sort.Sort(sorter)
+	if w.keepOrder {
+		task.rows = make([]chunk.Row, len(task.handles))
```
[the remark this answers is not in the document]
Comment (coocood) [reply executor/distsql.go]: We saved the orders in a map, we can directly put the row into the right position.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**